### PR TITLE
Implementation of Whois feature.

### PR DESCRIPTION
**Pull request by tchoutri**
### Summary of the changes
Hi @bitwalker! I have made the following changes in the code base:

* Implemented the relevant `handle_data` functions to catch the messages
* Added the numerics in `Irc.Commands`, except for `@rpl_whoismodes` that I have not encountered yet.
* Added a `whois_buffers` key in `%ClientState` to store the intermediate results.
* Added an `%Irc.Whois{}` struct to enforce the authorized keys and ensure data integrity and document the available fields.
* Updated the code to use the non-deprecated versions of `String.strip` and `String.to_char_list`.

### CODE REVIEW

**Comment by shymega** [comment]
@tchoutri This looks like a good PR, but I have just a few thoughts on it:

It would be even better with documentation: in the code itself, the examples directory (i.e, showing **how** to use the WHOIS function), and in the README perhaps, but the examples directory might cover that.. now I think about it.

The code itself looks good, but I'm not sure how to use it. I'll have another play when I get some more time. Good work though, although I'm not the maintainer, I've contributed before, and it certainly looks neater than my contributions :+1:

File: lib/exirc/client.ex [hunk 1]
```diff
@@ -589,6 +603,61 @@ defmodule ExIrc.Client do
...
+
+  ## WHOIS
+
+
+  def handle_data(%IrcMessage{cmd: @rpl_whoisuser, args: [_sender, nickname, username, hostname, _, realname]}, state) do
+    user = %{nickname: nickname, username: username, hostname: hostname, realname: realname}
+    {:noreply, %ClientState{state|whois_buffers: Map.put(state.whois_buffers, nickname, user)}}
+  end
+
+  def handle_data(%IrcMessage{cmd: @rpl_whoiscertfp, args: [_sender, nickname, "has client certificate fingerprint "<> fingerprint]}, state) do
```
Comment (bitwalker) [inline lib/exirc/client.ex]: Are we certain this text is standardized? It seems really fragile to match on it like this (though I'm not sure there is a better approach).

File: lib/exirc/client.ex [hunk 2]
```diff
@@ -589,6 +603,61 @@ defmodule ExIrc.Client do
...
+    {:noreply, %ClientState{state|whois_buffers: put_in(state.whois_buffers, [nickname, :certfp], fingerprint)}}
+  end
+
+  def handle_data(%IrcMessage{cmd: @rpl_whoisregnick, args: [_sender, nickname, _message]}, state) do
+    {:noreply, %ClientState{state|whois_buffers: put_in(state.whois_buffers, [nickname, :registered_nick?], true)}}
+  end
+
+  def handle_data(%IrcMessage{cmd: @rpl_whoishelpop, args: [_sender, nickname, "is available for help."]}, state) do
```
Comment (bitwalker) [inline lib/exirc/client.ex]: Same thing about the text matching

File: lib/exirc/client.ex [hunk 3]
```diff
@@ -589,6 +603,61 @@ defmodule ExIrc.Client do
...
+    {:noreply, %ClientState{state|whois_buffers: put_in(state.whois_buffers, [nickname, :helpop?], true)}}
+  end
+
+  def handle_data(%IrcMessage{cmd: @rpl_whoischannels, args: [_sender, nickname, channels]}, state) do
+    chans = String.split(channels, " ")
+    {:noreply, %ClientState{state|whois_buffers: put_in(state.whois_buffers, [nickname, :channels], chans)}}
+  end
+
+  def handle_data(%IrcMessage{cmd: @rpl_whoisserver, args: [_sender, nickname, server_addr, server_name]}, state) do
+    new_buffer = state.whois_buffers |> put_in([nickname, :server_name], server_name) |> put_in([nickname, :server_address], server_addr)
```
Comment (bitwalker) [inline lib/exirc/client.ex]: can we change the formatting here to:

```elixir
new_buffer =
  state.whois_buffers
  |> put_in(...)
  ...
```

**Comment by tchoutri** [comment]
Yes, in some cases, we will need to have the text to make the distinction between, for instance, `RPL_WHOISSADMIN` and ` RPL_NOTIFYACTION  ` (numeric 308), or `RPL_WHOISSADMIN ` and  `RPL_WHOISHELPER `. Look at [this up-to-date page about IRC Numerics](https://defs.ircdocs.horse/defs/numerics.html) for more informations.

However, after some reflexion, I suggest we deal with those on a case-by-case basis. :)

**Pull request merged**